### PR TITLE
[Fix] Loss average over non-ignored elements in CE loss

### PR DESCRIPTION
**Pull request by Wuziyi616**
Fix issue #546. I modify `CrossEntropyLoss` to be consistent with pytorch's official `F.cross_entropy`.

However, for Dice and Lovasz loss, I am not sure if we should also modify them accordingly since there is no such **official** implementation. I guess these two losses are common in medical image segmentation? What are the common practice of other papers?

### CODE REVIEW

**Comment by MengzhangLI** [comment]
Re-implementing TRAINING several experiments and here is one result.
`CONFIG=./configs/dmnet/dmnet_r101-d8_512x512_80k_ade20k.py`
`GPUS=4`
`GPUS_PER_NODE=4`
`CPUS_PER_TASK=8`

master branch:
+-------+-------+-------+
|  aAcc |  mIoU |  mAcc |
+-------+-------+-------+
| 81.16 | 45.13 | 56.34 |
+-------+-------+-------+

ziyi's branch:
+-------+-------+-------+
|  aAcc |  mIoU |  mAcc |
+-------+-------+-------+
| 81.09 | 44.47 | 56.16 |
+-------+-------+-------+

From https://github.com/open-mmlab/mmsegmentation/tree/master/configs/dmnet, reported mIoU is 45.42.

**Comment by MengzhangLI** [comment]
`CONFIG=./configs/dmnet/dmnet_r101-d8_512x512_80k_ade20k.py`
`GPUS=8`
`GPUS_PER_NODE=8`
`CPUS_PER_TASK=2`

master branch:
+-------+-------+-------+
|  aAcc |  mIoU |  mAcc |
+-------+-------+-------+
| 81.36 | 44.74 | 54.74 |
+-------+-------+-------+

ziyi's branch:
+-------+-------+-------+
|  aAcc |  mIoU |  mAcc |
+-------+-------+-------+
| 81.29 | 44.95 | 55.04 |
+-------+-------+-------+

From https://github.com/open-mmlab/mmsegmentation/tree/master/configs/dmnet, reported mIoU is 45.42.

**Comment by MengzhangLI** [comment]
`CONFIG=./configs/deeplabv3plus/deeplabv3plus_r101-d8_769x769_80k_cityscapes.py`
`GPUS=8`
`GPUS_PER_NODE=8`
`CPUS_PER_TASK=2`

master branch:
+---------------+-------+-------+
|     Class     |  IoU  |  Acc  |
+---------------+-------+-------+
|      road     | 98.28 | 98.98 |
|    sidewalk   | 85.88 | 93.18 |
|    building   | 93.09 | 96.94 |
|      wall     | 56.77 | 62.62 |
|     fence     |  65.2 | 73.66 |
|      pole     | 69.44 | 80.16 |
| traffic light | 75.42 | 86.46 |
|  traffic sign | 81.98 | 88.55 |
|   vegetation  | 92.79 | 96.69 |
|    terrain    | 63.46 | 73.82 |
|      sky      | 95.31 |  98.1 |
|     person    | 84.71 | 93.17 |
|     rider     | 67.53 | 80.03 |
|      car      | 95.92 | 98.18 |
|     truck     | 84.53 | 90.11 |
|      bus      | 87.66 |  94.7 |
|     train     | 70.75 | 74.24 |
|   motorcycle  |  70.6 | 83.04 |
|    bicycle    | 80.06 | 90.48 |
+---------------+-------+-------+
2021-07-11 14:20:38,799 - mmseg - INFO - Summary:
2021-07-11 14:20:38,800 - mmseg - INFO - 
+-------+-------+-------+
|  aAcc |  mIoU |  mAcc |
+-------+-------+-------+
| 96.43 | 79.97 | 87.01 |
+-------+-------+-------+



ziyi's branch:
+---------------+-------+-------+
|     Class     |  IoU  |  Acc  |
+---------------+-------+-------+
|      road     | 98.27 | 98.98 |
|    sidewalk   |  85.9 | 93.17 |
|    building   | 93.24 | 96.82 |
|      wall     | 55.25 | 62.13 |
|     fence     | 62.59 |  71.8 |
|      pole     | 69.78 | 80.75 |
| traffic light | 75.04 | 85.71 |
|  traffic sign | 82.17 | 88.13 |
|   vegetation  | 92.84 | 96.89 |
|    terrain    | 63.78 | 74.27 |
|      sky      | 95.49 | 98.07 |
|     person    | 84.57 | 93.15 |
|     rider     | 66.85 | 78.72 |
|      car      | 95.95 |  98.2 |
|     truck     | 79.41 | 91.41 |
|      bus      | 90.28 |  95.3 |
|     train     | 77.29 | 81.62 |
|   motorcycle  | 70.81 | 81.92 |
|    bicycle    | 80.47 | 90.03 |
+---------------+-------+-------+
2021-07-11 14:25:06,931 - mmseg - INFO - Summary:
2021-07-11 14:25:06,931 - mmseg - INFO - 
+-------+------+-------+
|  aAcc | mIoU |  mAcc |
+-------+------+-------+
| 96.43 | 80.0 | 87.21 |
+-------+------+-------+


From https://github.com/open-mmlab/mmsegmentation/tree/master/configs/dmnet, reported mIoU is 80.98

**Comment by Wuziyi616** [comment]
Thank you for your effort @MengzhangLI. Does that mean this PR is ok to merge? It seems to me that the performance is not affected.

**Comment by MengzhangLI** [comment]
> Thank you for your effort @MengzhangLI. Does that mean this PR is ok to merge? It seems to me that the performance is not affected.

Please check out latest results. Thanks! @xvjiarui @Junjun2016

**Comment by MengzhangLI** [comment]
`DATASET: Pascal VOC 2012 + Aug`


`CONFIG=./configs/deeplabv3plus/deeplabv3plus_r50-d8_512x512_40k_voc12aug.py`
`GPUS=4`
`GPUS_PER_NODE=4`
`CPUS_PER_TASK=2`

`ignore_index=255`

master branch:
+-------------+-------+-------+
|    Class    |  IoU  |  Acc  |
+-------------+-------+-------+
|  background |  94.3 | 97.65 |
|  aeroplane  | 91.47 | 95.77 |
|   bicycle   | 44.25 | 91.72 |
|     bird    | 87.83 | 92.58 |
|     boat    | 68.91 | 84.65 |
|    bottle   | 77.65 | 86.77 |
|     bus     | 95.33 | 97.65 |
|     car     | 87.45 | 93.49 |
|     cat     | 92.39 | 95.66 |
|    chair    | 37.02 |  49.0 |
|     cow     |  85.2 | 91.12 |
| diningtable | 60.08 | 64.36 |
|     dog     | 86.45 | 92.94 |
|    horse    | 86.45 | 92.52 |
|  motorbike  | 86.85 | 91.68 |
|    person   | 86.13 |  91.6 |
| pottedplant | 58.11 |  70.7 |
|    sheep    | 84.31 | 91.39 |
|     sofa    | 46.35 | 52.92 |
|    train    | 91.19 |  95.2 |
|  tvmonitor  | 69.16 | 79.39 |
+-------------+-------+-------+
+-------+------+-------+
|  aAcc | mIoU |  mAcc |
+-------+------+-------+
| 94.89 | 77.0 | 85.66 |
+-------+------+-------+

ziyi's branch:
+-------------+-------+-------+
|    Class    |  IoU  |  Acc  |
+-------------+-------+-------+
|  background | 94.46 | 97.66 |
|  aeroplane  | 90.51 | 95.08 |
|   bicycle   |  43.9 | 90.95 |
|     bird    | 86.01 | 91.81 |
|     boat    | 67.79 | 87.76 |
|    bottle   | 75.35 | 81.55 |
|     bus     |  95.1 | 97.82 |
|     car     | 87.47 | 92.98 |
|     cat     | 92.02 | 96.24 |
|    chair    | 38.25 | 53.11 |
|     cow     | 90.06 |  93.1 |
| diningtable | 52.09 | 55.89 |
|     dog     | 84.76 | 92.15 |
|    horse    | 88.99 | 94.34 |
|  motorbike  | 85.78 | 90.81 |
|    person   | 85.75 | 92.24 |
| pottedplant | 62.31 | 75.88 |
|    sheep    | 85.01 | 90.96 |
|     sofa    | 48.73 | 56.63 |
|    train    | 90.68 | 93.22 |
|  tvmonitor  | 72.05 | 79.15 |
+-------------+-------+-------+
+------+------+-------+
| aAcc | mIoU |  mAcc |
+------+------+-------+
| 94.9 | 77.0 | 85.68 |
+------+------+-------+

From https://github.com/open-mmlab/mmsegmentation/tree/master/configs/deeplabv3plus, reported mIoU is 76.81

**Comment by MengzhangLI** [comment]
`DATASET: Pascal VOC 2012 + Aug`

`CONFIG=./configs/pspnet/pspnet_r50-d8_512x512_40k_voc12aug.py`
`GPUS=4`
`GPUS_PER_NODE=4`
`CPUS_PER_TASK=2`

`ignore_index=255`

master branch:
+-------------+-------+-------+
|    Class    |  IoU  |  Acc  |
+-------------+-------+-------+
|  background | 94.37 | 97.56 |
|  aeroplane  | 91.81 | 95.99 |
|   bicycle   | 43.98 |  90.4 |
|     bird    | 87.31 | 93.83 |
|     boat    | 71.68 | 87.72 |
|    bottle   | 76.23 | 84.05 |
|     bus     | 95.34 | 97.56 |
|     car     | 87.05 | 93.65 |
|     cat     | 92.93 | 96.29 |
|    chair    | 38.06 | 52.12 |
|     cow     | 89.85 | 93.48 |
| diningtable | 59.06 | 62.03 |
|     dog     | 87.21 | 95.72 |
|    horse    | 88.82 | 94.68 |
|  motorbike  | 86.81 | 91.68 |
|    person   | 85.35 | 92.38 |
| pottedplant | 61.07 | 72.17 |
|    sheep    | 84.14 | 88.81 |
|     sofa    | 47.25 | 53.47 |
|    train    | 88.75 | 92.76 |
|  tvmonitor  | 69.69 | 78.62 |
+-------------+-------+-------+
+-------+-------+-------+
|  aAcc |  mIoU |  mAcc |
+-------+-------+-------+
| 94.94 | 77.46 | 85.95 |
+-------+-------+-------+

ziyi's branch:
+-------------+-------+-------+
|    Class    |  IoU  |  Acc  |
+-------------+-------+-------+
|  background | 94.31 | 97.75 |
|  aeroplane  | 90.59 | 95.95 |
|   bicycle   | 44.05 | 89.85 |
|     bird    | 86.62 | 92.09 |
|     boat    | 70.16 | 85.33 |
|    bottle   | 71.65 | 79.85 |
|     bus     | 95.25 | 97.29 |
|     car     |  85.9 | 92.47 |
|     cat     | 91.57 |  95.7 |
|    chair    | 38.59 | 50.99 |
|     cow     | 89.51 | 93.83 |
| diningtable | 56.17 | 58.36 |
|     dog     | 86.98 | 94.28 |
|    horse    | 89.61 | 94.52 |
|  motorbike  | 85.06 | 90.65 |
|    person   | 85.07 | 91.92 |
| pottedplant | 63.95 | 74.06 |
|    sheep    | 82.87 | 89.38 |
|     sofa    |  46.6 |  51.8 |
|    train    | 90.97 |  94.1 |
|  tvmonitor  | 73.87 |  79.7 |
+-------------+-------+-------+
+-------+-------+-------+
|  aAcc |  mIoU |  mAcc |
+-------+-------+-------+
| 94.89 | 77.11 | 85.23 |
+-------+-------+-------+


From https://github.com/MengzhangLI/mmsegmentation/tree/master/configs/pspnet, reported mIoU is 77.29

File: mmseg/models/losses/cross_entropy_loss.py [hunk 1]
```diff
@@ -24,6 +24,11 @@ def cross_entropy(pred,
         ignore_index=ignore_index)
 
     # apply weights and do the reduction
+    # pytorch's official cross_entropy average loss over non-ignored elements
+    # refer to https://github.com/pytorch/pytorch/blob/56b43f4fec1f76953f15a627694d4bba34588969/torch/nn/functional.py#L2660  # noqa
+    if avg_factor is None:
+        avg_factor = label.numel() - (label == ignore_index).sum().item()
```
Comment (sunyc11) [inline mmseg/models/losses/cross_entropy_loss.py:30]: the `avg_factor` will be zero if all elements in `label` is `ignore_index`, and in `weight_reduce_loss` function, `loss.sum() / avg_factor` will cause `nan`

Comment (Wuziyi616) [reply mmseg/models/losses/cross_entropy_loss.py:30]: Hi, thanks for reviewing the PR. Since I'm no longer at SenseTime, please feel free to close this PR and recreate a fix. Thanks
